### PR TITLE
Optimize role bitset in document level security

### PR DESCRIPTION
**Pull request by dnhatn**
If a role query matches all documents in a leaf reader, then we can use a special bitset to reduce the memory usage and provide a faster path to calculate the num docs and the combined liveDocs.

### CODE REVIEW

**Review by tvernum — COMMENTED** [review]
The core change seems good to me, but it feels like the test is over relying on randomisation.

I would prefer explicit tests for small number matches (sparse bitset), large number of matches (fixed bitset) & all matched (matchall bitset).

**Comment by dnhatn** [comment]
> it feels like the test is over relying on randomisation.

I pushed https://github.com/elastic/elasticsearch/pull/64673/commits/45de38b503695aab67c50c5995b966ef695d33d1  to separate that test to three smaller tests. Thanks for looking.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/accesscontrol/DocumentSubsetBitsetCache.java [hunk 1]
```diff
@@ -339,4 +342,30 @@ void verifyInternalConsistency() {
...
+        }
+        if (iter.docID() != -1) {
+            throw new IllegalStateException("Must be an unpositioned iterator, got current position = " + iter.docID());
+        }
+        int matches = 0;
+        for (int doc = iter.nextDoc(); doc != DocIdSetIterator.NO_MORE_DOCS; doc = iter.nextDoc()) {
+            matches++;
+            set.set(doc);
+        }
+        if (matches == maxDoc) {
```
Comment (jpountz) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/accesscontrol/DocumentSubsetBitsetCache.java:364]: The BitSet#or implementations seem to have some optimizations, maybe we should do set.or(it) and then compare set.cardinality() with maxDoc.

**Review by jpountz — APPROVED** [review]

**Review by jpountz — APPROVED** [review]

**Comment by dnhatn** [comment]
run elasticsearch-ci/2

**Review by tvernum — APPROVED** [review]
LGTM

**Comment by dnhatn** [comment]
@tvernum @jpountz Thanks for reviewing.

**Pull request merged**